### PR TITLE
client: add MetaSessionRef/MetaRequestRef and MetaRequest::request_lock support

### PR DESCRIPTION
**Pull request by lxbsz**
The `request_lock` and `client_lock` locking order should always be:

request_lock.lock();
...
client_lock.lock();
...
client_lock.unlock();
...
request_lock.unlock();

And also make the MetaRequest:_ref = {0} as default, when creating
the MetaRequest object, it will hold by the MetaRequestRef to simplify
the code. And when adding it to mds_requests map, it will increase
the _ref and it will be decreased by unregister_request(), the usage
will be:

func() {
  MetaRequestRef req = new MetaRequest(this, ...);
  make_request(req.get(), ...);
}

In any other functions when it want to access to the request in
mds_reqeusts map, it should do:

func() {
  client_lock.lock();
  MetaRequestRef req = mds_requests.at(tid);
  client_lock.unlock();

  req->request_lock.lock();
  req->xyz...
  req->request_lock.unlock();
}

Fixes: https://tracker.ceph.com/issues/46890
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/client/MetaRequest.h [hunk 1]
```diff
@@ -47,12 +53,12 @@ struct MetaRequest {
   __u32    sent_on_mseq;       // mseq at last submission of this request
   int      num_fwd;            // # of times i've been forwarded
   int      retry_attempt;
-  std::atomic<uint64_t> ref = { 1 };
-  
+  std::atomic<uint64_t> ref = { 0 };
```
Comment (batrick) [inline src/client/MetaRequest.h:56]: When building ref counted objects, it's better to inherit from `RefCountedObject` in `src/common`.

Comment (lxbsz) [reply src/client/MetaRequest.h:56]: Sure, I will do that.

Comment (lxbsz) [reply src/client/MetaRequest.h]: Done.

**Comment by lxbsz** [comment]
> Better would be to use `std::shared_ptr` for both `MetaRequest` and `MetaSession`. Then use a `std::weak_ptr` in `Client::mds_sessions` (with checks at time of use and trimming).
> 
> For `MetaRequest`, you can build a queue for trimming the inode that's processed in `Client::tick`.
> 
> I think that should work much better?

Sorry, I missed this one, will have a check of it.

**Comment by lxbsz** [comment]
> Better would be to use `std::shared_ptr` for both `MetaRequest` and `MetaSession`. Then use a `std::weak_ptr` in `Client::mds_sessions` (with checks at time of use and trimming).
> 
> For `MetaRequest`, you can build a queue for trimming the inode that's processed in `Client::tick`.
> 
> I think that should work much better?

Yeah, for the `MetaSession` it will work well with the `shared_ptr`. But for the `MetaRequest`, we need to `try_to_trim_inode(other_in.get(), false); ` at last just when freeing the request memory. So we need an intrusive ptr to do this in the destructor.

**Comment by lxbsz** [comment]
Updated it. Thanks.

**Comment by lxbsz** [comment]
Updated it by switching to use delay put requests, which is the same with https://github.com/ceph/ceph/pull/36580, and also it will depend on the commit https://github.com/ceph/ceph/pull/36580/commits/e989f7f6646663298f9bd4f63337d713b5c00a72 in https://github.com/ceph/ceph/pull/36580/files.

**Comment by lxbsz** [comment]
Renamed the `requests_delay_release` --> `delay_r_release`.

**Comment by lxbsz** [comment]
Resolved the confilicts.

**Comment by lxbsz** [comment]
jenkins test make check

**Comment by lxbsz** [comment]
jenkins retest this please

File: src/client/Client.cc [hunk 2]
```diff
@@ -3534,25 +3534,20 @@ static int adjust_caps_used_for_lazyio(int used, int issued, int implemented)
   if (!(implemented & CEPH_CAP_FILE_LAZYIO))
     return used;
 
+  used |= CEPH_CAP_FILE_LAZYIO;
```
Comment (jtlayton) [inline src/client/Client.cc:3669]: Isn't this changing the logic in this function? Before, LAZYIO would only be set in the returned flags iff any of Fcbl were issued. Now, it's always returned even if none of them are issued.

Comment (lxbsz) [reply src/client/Client.cc]: Fixed it. Thanks.

**Comment by batrick** [comment]
rebase please

**Comment by lxbsz** [comment]
@batrick Rebased to the latest, PTAL, thanks.

**Comment by lxbsz** [comment]
Rebased it.

**Comment by lxbsz** [comment]
Rebased.

**Comment by lxbsz** [comment]
jenkins test api

**Comment by lxbsz** [comment]
jenkins test api

**Comment by lxbsz** [comment]
This series has been folded into https://github.com/ceph/ceph/pull/39797, closing this one.

**Pull request closed without merging**